File: src/engine/export-profile.ts
import { z } from "zod";

import { POIETRA_ENGINE_CONTRACT_VERSION } from "./primitives";

/**
 * Ceiling for an export profile's declared duration bound, in seconds.
 *
 * Matches the 900-second (15-minute) ceiling of the historically accepted
 * server render profile (`MAX_VIDEO_DURATION_SECONDS = 15 * 60` in
 * `sandbox/manim-render-gated-oci/render-entrypoint.py`).
 */
export const MAX_EXPORT_DURATION_SECONDS = 900;

/**
 * Ceiling for an export profile's declared output-size bound, in bytes (128 MiB).
 *
 * Matches `MAX_ARTIFACT_BYTES = 128 * 1024 * 1024` in
 * `sandbox/manim-render-gated-oci/render-entrypoint.py`.
 */
export const MAX_EXPORT_OUTPUT_BYTES = 134_217_728;

/**
 * Closed 16:9 output ladder aligned to the 8 x (128/9) scene-unit frame.
 *
 * 854x480 preserves the legacy accepted server render profile (854 px is the
 * even-width rounding of 480 x 16/9); 1280x720 and 1920x1080 are the exact
 * 16:9 rungs above it. Arbitrary integer dimensions are rejected by
 * construction.
 */
export const exportResolutionV1Schema = z.enum(["854x480", "1280x720", "1920x1080"]);

export type ExportResolutionV1 = z.infer<typeof exportResolutionV1Schema>;

export const EXPORT_RESOLUTION_PIXELS_V1 = {
  "854x480": { heightPx: 480, widthPx: 854 },
  "1280x720": { heightPx: 720, widthPx: 1280 },
  "1920x1080": { heightPx: 1080, widthPx: 1920 },
} as const satisfies Record<ExportResolutionV1, Readonly<{ heightPx: number; widthPx: number }>>;

/** Closed v1 sampling/output frame rate: the integer 30 or 60. */
export const exportFrameRateV1Schema = z.union([z.literal(30), z.literal(60)]);

/**
 * Closed v1 description of one client-side video export.
 *
 * Sampling frame-rate precedence (#693 Question 6 decision): the export
 * profile supplies the sampling fps. Export stepping walks output frame
 * indices `i` and samples the Scene at `state_sample_time(i / frameRate)`
 * using this profile's `frameRate`. `SceneIrV1.stateSampling.frameRate`
 * remains the Scene's own declaration of its retained-state sampling grid and
 * is NOT overwritten by the profile: the profile decides which instants are
 * requested, and the Scene's `stateSampling` decides how each requested
 * instant resolves to retained state. A Scene declaring `frameRate: null`
 * samples continuously at exactly the profile's requested instants.
 */
export const exportProfileV1Schema = z
  .object({
    codec: z.literal("h264-mp4"),
    colorContractVersion: z.literal(1),
    frameRate: exportFrameRateV1Schema,
    maxDurationSeconds: z.number().int().positive().max(MAX_EXPORT_DURATION_SECONDS),
    maxOutputBytes: z.number().int().positive().max(MAX_EXPORT_OUTPUT_BYTES),
    resolution: exportResolutionV1Schema,
    schema: z.literal("poietra.export-profile"),
    version: z.literal(POIETRA_ENGINE_CONTRACT_VERSION),
  })
  .strict();

export type ExportProfileV1 = z.infer<typeof exportProfileV1Schema>;

/**
 * Every profile field is a closed literal or an in-schema bound, so this Zod
 * parse is the complete v1 admission and mirrors the Rust
 * `parse_export_profile_json_v1`.
 */
export function parseExportProfileV1(value: unknown): ExportProfileV1 {
  return exportProfileV1Schema.parse(value);
}

function canonicalProfileMetadata(profile: ExportProfileV1) {
  return {
    codec: profile.codec,
    colorContractVersion: profile.colorContractVersion,
    frameRate: profile.frameRate,
    maxDurationSeconds: profile.maxDurationSeconds,
    maxOutputBytes: profile.maxOutputBytes,
    resolution: profile.resolution,
    schema: profile.schema,
    version: profile.version,
  };
}

/** Returns the byte-for-byte canonical profile JSON shared with the Rust v1 contract. */
export function canonicalExportProfileV1(profile: ExportProfileV1) {
  return JSON.stringify(canonicalProfileMetadata(profile));
}

/**
 * Computes the lower-case SHA-256 digest over the canonical export profile.
 *
 * The profile does not embed its own digest; this value names a profile in
 * publication lineage.
 */
export async function digestExportProfileV1(profile: ExportProfileV1) {
  const bytes = new TextEncoder().encode(canonicalExportProfileV1(profile));
  const digest = await globalThis.crypto.subtle.digest("SHA-256", bytes);
  return Array.from(new Uint8Array(digest), (byte) => byte.toString(16).padStart(2, "0")).join("");
}

File: engine/crates/poietra-scene-ir/src/export_profile.rs
//! `ExportProfileV1` — the closed v1 contract describing one client-side video export.
//!
//! The profile is the single authority for *how* a Scene is encoded — output
//! resolution, sampling frame rate, codec tier, color contract version, and the
//! declared duration/output-size bounds — and it never modifies the Scene
//! itself. Consumers (export sessions, muxers, publication acceptance) are
//! introduced separately; this module only defines and validates the contract.

use serde::{Deserialize, Deserializer, Serialize, Serializer, de};

use crate::model::{ContractVersionV1, deserialize_js_safe_u32, deserialize_js_safe_u64};

#[derive(Clone, Copy, Debug, Default, Eq, PartialEq, Serialize, Deserialize)]
pub enum ExportProfileSchemaV1 {
    #[default]
    #[serde(rename = "poietra.export-profile")]
    ExportProfile,
}

/// The only codec tier accepted by v1: H.264 video in an MP4 container.
///
/// H.264/MP4 is the most widely supported `WebCodecs` encode target and the
/// only container/codec pair the existing server acceptance profile
/// (`sandbox/manim-render-gated-oci/render-entrypoint.py::_validate_mp4`) has
/// ever admitted. Additional tiers (AV1/VP9 + `WebM`) are a later, opt-in
/// phase.
#[derive(Clone, Copy, Debug, Default, Eq, PartialEq, Serialize, Deserialize)]
pub enum ExportCodecTierV1 {
    #[default]
    #[serde(rename = "h264-mp4")]
    H264Mp4,
}

/// The only export color contract version accepted by v1.
///
/// The integer version-locks the export color pipeline without restating it:
/// version 1 names the contract under which export frames are sRGB-encoded
/// RGBA readback. The normative definition (encoder tagging, `colr` boxes,
/// roundtrip gates) belongs to the export-authority color ADR.
#[derive(Clone, Copy, Debug, Default, Eq, PartialEq)]
pub struct ExportColorContractVersionV1;

impl Serialize for ExportColorContractVersionV1 {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        serializer.serialize_u8(1)
    }
}

impl<'de> Deserialize<'de> for ExportColorContractVersionV1 {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        let version = deserialize_js_safe_u64(deserializer)?;
        if version == 1 {
            Ok(Self)
        } else {
            Err(de::Error::custom(format!(
                "unsupported export color contract version {version}; expected 1"
            )))
        }
    }
}

/// Closed 16:9 output ladder aligned to the 8 × (128/9) scene-unit frame.
///
/// 1280×720 and 1920×1080 are exact 16:9 pixel grids: they rasterize the same
/// camera frame at different densities and never change what is visible.
/// 854×480 preserves the legacy accepted server render profile — 854 px is
/// the even-width rounding of 480 × 16/9 — so its pixel aspect deviates from
/// 16:9 by less than one pixel column; export sampling resolves that
/// deviation exactly as the legacy Manim profile did, by widening the camera
/// window sub-pixel about its center (never cropping). Arbitrary integer
/// dimensions are rejected by construction: the wire value is one of these
/// three literals.
#[derive(Clone, Copy, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub enum ExportResolutionV1 {
    #[serde(rename = "854x480")]
    Sd854x480,
    #[serde(rename = "1280x720")]
    Hd1280x720,
    #[serde(rename = "1920x1080")]
    FullHd1920x1080,
}

impl ExportResolutionV1 {
    #[must_use]
    pub const fn width_px(&self) -> u32 {
        match self {
            Self::Sd854x480 => 854,
            Self::Hd1280x720 => 1280,
            Self::FullHd1920x1080 => 1920,
        }
    }

    #[must_use]
    pub const fn height_px(&self) -> u32 {
        match self {
            Self::Sd854x480 => 480,
            Self::Hd1280x720 => 720,
            Self::FullHd1920x1080 => 1080,
        }
    }
}

/// Closed v1 sampling/output frame rate, carried on the wire as the integer
/// 30 or 60.
#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum ExportFrameRateV1 {
    Fps30,
    Fps60,
}

impl ExportFrameRateV1 {
    #[must_use]
    pub const fn frames_per_second(&self) -> f64 {
        match self {
            Self::Fps30 => 30.0,
            Self::Fps60 => 60.0,
        }
    }
}

impl Serialize for ExportFrameRateV1 {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        serializer.serialize_u8(match self {
            Self::Fps30 => 30,
            Self::Fps60 => 60,
        })
    }
}

impl<'de> Deserialize<'de> for ExportFrameRateV1 {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        match deserialize_js_safe_u64(deserializer)? {
            30 => Ok(Self::Fps30),
            60 => Ok(Self::Fps60),
            value => Err(de::Error::custom(format!(
                "unsupported export frame rate {value}; expected 30 or 60"
            ))),
        }
    }
}

/// Closed v1 description of one client-side video export.
///
/// # Sampling frame-rate precedence (#693 Question 6 decision)
///
/// The export profile supplies the sampling fps. Export stepping walks output
/// frame indices `i` and samples the Scene at
/// [`crate::SceneIrV1::state_sample_time`]`(i / frameRate)` using this
/// profile's `frame_rate`. `SceneIrV1.state_sampling.frame_rate` remains the
/// Scene's own declaration of its retained-state sampling grid and is NOT
/// overwritten by the profile. The precedence rule is composition, not
/// override:
///
/// - the profile decides *which instants are requested* — `i / frameRate` for
///   each output frame index `i`;
/// - the Scene's `state_sampling` decides *how each requested instant
///   resolves* to retained state — `state_sample_time` quantizes onto the
///   Scene's own `frame_rate` grid when that declaration is non-null and
///   applies the terminal-retention rule at the endpoint.
///
/// An exporter therefore never writes to `state_sampling.frame_rate`; a Scene
/// declaring `frame_rate: null` samples continuously at exactly the profile's
/// requested instants.
#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct ExportProfileV1 {
    pub codec: ExportCodecTierV1,
    pub color_contract_version: ExportColorContractVersionV1,
    /// Output and sampling frame rate. See the precedence rule on
    /// [`ExportProfileV1`].
    pub frame_rate: ExportFrameRateV1,
    /// Declared upper bound on encoded video duration, in whole seconds.
    ///
    /// Must stay within [`crate::MAX_EXPORT_DURATION_SECONDS_V1`].
    #[serde(deserialize_with = "deserialize_js_safe_u32")]
    pub max_duration_seconds: u32,
    /// Declared upper bound on the finalized container size, in bytes.
    ///
    /// Must stay within [`crate::MAX_EXPORT_OUTPUT_BYTES_V1`].
    #[serde(deserialize_with = "deserialize_js_safe_u64")]
    pub max_output_bytes: u64,
    pub resolution: ExportResolutionV1,
    pub schema: ExportProfileSchemaV1,
    pub version: ContractVersionV1,
}

#[cfg(test)]
mod export_wire_tests {
    use super::*;

    #[test]
    fn frame_rate_matches_javascript_number_semantics() {
        assert_eq!(
            serde_json::from_str::<ExportFrameRateV1>("30.0").unwrap(),
            ExportFrameRateV1::Fps30
        );
        assert_eq!(
            serde_json::from_str::<ExportFrameRateV1>("6e1").unwrap(),
            ExportFrameRateV1::Fps60
        );
        assert!(serde_json::from_str::<ExportFrameRateV1>("59.94").is_err());
        assert!(serde_json::from_str::<ExportFrameRateV1>("24").is_err());
    }

    #[test]
    fn color_contract_version_is_closed_at_one() {
        assert_eq!(
            serde_json::from_str::<ExportColorContractVersionV1>("1.0").unwrap(),
            ExportColorContractVersionV1
        );
        assert!(serde_json::from_str::<ExportColorContractVersionV1>("2").is_err());
        assert_eq!(
            serde_json::to_string(&ExportColorContractVersionV1).unwrap(),
            "1"
        );
    }

    #[test]
    fn resolution_ladder_is_closed_and_carries_pixel_dimensions() {
        let ladder = [
            (ExportResolutionV1::Sd854x480, "\"854x480\"", 854, 480),
            (ExportResolutionV1::Hd1280x720, "\"1280x720\"", 1280, 720),
            (
                ExportResolutionV1::FullHd1920x1080,
                "\"1920x1080\"",
                1920,
                1080,
            ),
        ];
        for (resolution, wire, width, height) in ladder {
            assert_eq!(serde_json::to_string(&resolution).unwrap(), wire);
            assert_eq!(
                serde_json::from_str::<ExportResolutionV1>(wire).unwrap(),
                resolution
            );
            assert_eq!(resolution.width_px(), width);
            assert_eq!(resolution.height_px(), height);
        }
        assert!(serde_json::from_str::<ExportResolutionV1>("\"640x360\"").is_err());
        assert!(serde_json::from_str::<ExportResolutionV1>("\"3840x2160\"").is_err());
    }
}
